File: codes/vimkim/3_longestsubstringwithoutrepeatingcharacters.rs
use std::collections::HashSet;

impl Solution {

    fn length_of_longest_substring(s: String) -> i32 {
        let mut set = HashSet::new();
        let s_bytes = s.as_bytes();
        let mut l = 0;
        let mut r = 0;
        let mut ans = 0;
        let n = s_bytes.len();

        while r < n {
            if !set.contains(&s_bytes[r]) {
              // copy
                set.insert(s_bytes[r]);
                r += 1;
                ans = ans.max((r - l) as i32);
            } else {
                set.remove(&s_bytes[l]);
                l += 1;
            }
        }
        ans
    }

}

File: codes/vimkim/3_longestsubstringwithoutrepeatingcharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        set_ = set()

        l = r = 0
        N = len(s)
        ans = 0

        while r < N:
            if s[r] not in set_:
                set_.add(s[r])
                r += 1
                ans = max(ans, r - l) # length of peaceful period 무사고 기간

            else: # dance
                set_.remove(s[l])
                l += 1 # day 1 -> 2 -> 3 ...

                49th 50 - 99

        return ans